js/src/vm/PortableBaselineInterpret.cpp

```diff
@@ -1412,6 +1431,59 @@ ICInterpretOps(BaselineFrame* frame, VMFrameManager& frameMgr, State& state,
     DISPATCH_CACHEOP();
   }
 
+  CACHEOP_CASE(StoreTypedArrayElement) {
+    ObjOperandId objId = icregs.cacheIRReader.objOperandId();
+    Scalar::Type elementType = icregs.cacheIRReader.scalarType();
+    IntPtrOperandId indexId = icregs.cacheIRReader.intPtrOperandId();
+    uint32_t rhsId = icregs.cacheIRReader.rawOperandId();
+    bool handleOOB = icregs.cacheIRReader.readBool();
+    JSObject* obj = reinterpret_cast<JSObject*>(icregs.icVals[objId.id()]);
+    uintptr_t index = uintptr_t(icregs.icVals[indexId.id()]);
+    uint64_t rhs = icregs.icVals[rhsId];
+    if (index >= obj->as<TypedArrayObject>().length()) {
+      if (!handleOOB) {
+        return ICInterpretOpResult::NextIC;
+      }
+    } else {
+      Value v;
+      switch (elementType) {
+        case Scalar::Int8:
+        case Scalar::Uint8:
+        case Scalar::Int16:
+        case Scalar::Uint16:
+        case Scalar::Int32:
+        case Scalar::Uint32:
+        case Scalar::Uint8Clamped:
+          v = Int32Value(rhs);
+          break;
+
+        case Scalar::Float32:
+        case Scalar::Float64:
+          v = Value::fromRawBits(rhs);
+          break;
+
+        case Scalar::BigInt64:
+        case Scalar::BigUint64:
+          v = BigIntValue(reinterpret_cast<JS::BigInt*>(rhs));
+          break;
+
+        case Scalar::MaxTypedArrayViewType:
+        case Scalar::Int64:
+        case Scalar::Simd128:
+          MOZ_CRASH("Unsupported TypedArray type");
+      }
+
+      ReservedRooted<JSObject*> obj0(&state.obj0, obj);
+      ReservedRooted<Value> value0(&state.value0, v);
+      PUSH_IC_FRAME();
+      ObjectOpResult result;
+      MOZ_ALWAYS_TRUE(SetTypedArrayElement(cx, obj0.as<TypedArrayObject>(),
```

Comment:
I'm curious if we can dig a little further on whether `cx` is needed here -- it's pretty unfortunate to push an IC frame for every array-set in a tight inner loop! It looks like `cx->resultToBool()` is used in places; but otherwise maybe we don't have slowpath conversions that allocate or anything, since we have strong type guarantees from the setup of `v` above? If we can work this out and leave a comment, and use `frame.cxForLocalUseOnly()`, that would be ideal; if that seems to hard to justify though, no worries, I just want to push a little more on it first.

Reply:
Okay, I've dug into this some more and I agree that it looks like we do have enough guarantees about the types of `Value`s we may store that the exit frame shouldn't be necessary. I'm not sure I can prove it with any confidence though. There's enough template overloading and macros that I don't even know if I found all the relevant code.

So I tested this by just passing `nullptr` for `cx` and seeing what fails. The only thing which goes wrong is that for `BigInt` values, `ToBigInt` stores the `HandleValue` it receives into a `RootedValue`, which needs a valid context.

Passing `frameMgr.cxForLocalUseOnly()` instead does pass all the jit-tests. So does using a `FakeRooted<JSObject*>` instead of `ReservedRooted`. I couldn't figure out how to get a `Handle<Value>` from a `FakeRooted<Value>` though, so I'm still using `ReservedRooted` for that.

Reply:
Awesome, thanks for digging into that!